### PR TITLE
Use a "get_bool" function for boolean args in rest

### LINKED ISSUE #8079: Make boolean values consistent across REST API
```
.... according to the [OpenAPI specification](https://swagger.io/docs/specification/v3_0/data-models/data-types/#boolean) for boolean types:

> type: boolean represents two values: true and false. Note that truthy and 
> falsy values such as "true", "", 0 or null are not considered boolean values.

While accepting '1' and '0' maintains consistency with some existing endpoints, 
the OpenAPI spec is clear that booleans should be 'true' or 'false'. Since this 
is a new parameter, we could argue for strict spec compliance here.
```

_Originally posted by @mgajek-cern in https://github.com/rucio/rucio/pull/8072#discussion_r2393731645_

### PR DESCRIPTION
**Pull request by voetberg**
Closes #8079

### CODE REVIEW

File: lib/rucio/web/rest/flaskapi/v1/replicas.py [hunk 1]
```diff
@@ -361,7 +361,7 @@ def post(self):
                 files=files,
                 issuer=request.environ['issuer'],
                 vo=request.environ['vo'],
-                ignore_availability=param_get(parameters, 'ignore_availability', default=False),
+                ignore_availability=bool(param_get_bool(parameters, 'ignore_availability', default=False)),
```
Comment (mgajek-cern) [inline lib/rucio/web/rest/flaskapi/v1/replicas.py]: bool() wrapper might be redundant here. `param_get_bool` already returns a bool

File: lib/rucio/web/rest/flaskapi/v1/common.py [hunk 2]
```diff
@@ -425,6 +425,45 @@ def param_get(parameters: dict[str, Any], name: str, **kwargs) -> Any:
         return parameters[name]
 
 
+def param_get_bool(parameters: dict[str, Any], name: str, **kwargs) -> bool:
```
Comment (mgajek-cern) [inline lib/rucio/web/rest/flaskapi/v1/common.py]: You may want to reuse the `param_get_bool` function in other parts of the code:

```
grep -r "default=False" 
./flaskapi/v1/dids.py:            ignore_duplicate = param_get(parameters, 'ignore_duplicate', default=False)
./flaskapi/v1/dids.py:        inherit = param_get(params, "inherit", default=False)
./flaskapi/v1/replicas.py:                ignore_availability=param_get(parameters, 'ignore_availability', default=False),
./flaskapi/v1/replicas.py:        unavailable = param_get(parameters, 'unavailable', default=False)
./flaskapi/v1/replicas.py:        all_states = param_get(parameters, 'all_states', default=False)
./flaskapi/v1/replicas.py:        resolve_parents = param_get(parameters, 'resolve_parents', default=False)
./flaskapi/v1/replicas.py:        force = param_get(parameters, 'force', default=False)
./flaskapi/v1/replicas.py:            deep = request.args.get('deep', default=False)
./flaskapi/v1/accounts.py:                default=param_get(parameters, 'default', default=False),
./flaskapi/v1/vos.py:        default = param_get(parameters, 'default', default=False)
./flaskapi/v1/subscriptions.py:        dry_run = param_get(options, 'dry_run', default=False)
./flaskapi/v1/subscriptions.py:        priority = param_get(options, 'priority', default=False)
./flaskapi/v1/rules.py:        estimate_ttc = param_get(parameters, 'estimate_ttc', default=False)
./flaskapi/v1/rules.py:                locked=param_get(parameters, 'locked', default=False),
./flaskapi/v1/rules.py:                purge_replicas=param_get(parameters, 'purge_replicas', default=False),
./flaskapi/v1/rules.py:                ignore_availability=param_get(parameters, 'ignore_availability', default=False),
./flaskapi/v1/rules.py:                ask_approval=param_get(parameters, 'ask_approval', default=False),
./flaskapi/v1/rules.py:                asynchronous=param_get(parameters, 'asynchronous', default=False),
./flaskapi/v1/rules.py:                split_container=param_get(parameters, 'split_container', default=False),
marvingajek@Marvins-MacBook-Pro rest % grep -r "default=False" 
./flaskapi/v1/dids.py:            ignore_duplicate = param_get(parameters, 'ignore_duplicate', default=False)
./flaskapi/v1/dids.py:        inherit = param_get(params, "inherit", default=False)
./flaskapi/v1/replicas.py:                ignore_availability=param_get(parameters, 'ignore_availability', default=False),
./flaskapi/v1/replicas.py:        unavailable = param_get(parameters, 'unavailable', default=False)
./flaskapi/v1/replicas.py:        all_states = param_get(parameters, 'all_states', default=False)
./flaskapi/v1/replicas.py:        resolve_parents = param_get(parameters, 'resolve_parents', default=False)
./flaskapi/v1/replicas.py:        force = param_get(parameters, 'force', default=False)
./flaskapi/v1/replicas.py:            deep = request.args.get('deep', default=False)
./flaskapi/v1/accounts.py:                default=param_get(parameters, 'default', default=False),
./flaskapi/v1/vos.py:        default = param_get(parameters, 'default', default=False)
./flaskapi/v1/subscriptions.py:        dry_run = param_get(options, 'dry_run', default=False)
./flaskapi/v1/subscriptions.py:        priority = param_get(options, 'priority', default=False)
./flaskapi/v1/rules.py:        estimate_ttc = param_get(parameters, 'estimate_ttc', default=False)
./flaskapi/v1/rules.py:                locked=param_get(parameters, 'locked', default=False),
./flaskapi/v1/rules.py:                purge_replicas=param_get(parameters, 'purge_replicas', default=False),
./flaskapi/v1/rules.py:                ignore_availability=param_get(parameters, 'ignore_availability', default=False),
./flaskapi/v1/rules.py:                ask_approval=param_get(parameters, 'ask_approval', default=False),
./flaskapi/v1/rules.py:                asynchronous=param_get(parameters, 'asynchronous', default=False),
./flaskapi/v1/rules.py:                split_container=param_get(parameters, 'split_container', default=False),
marvingajek@Marvins-MacBook-Pro rest % grep -r "default=True"  
./flaskapi/v1/replicas.py:        resolve_archives = param_get(parameters, 'resolve_archives', default=True)
marvingajek@Marvins-MacBook-Pro rest % 
```

Comment (mgajek-cern) [inline lib/rucio/web/rest/flaskapi/v1/common.py]: Would it make sense to add unit tests for `common.py` and the other files in the `rest` folder to ensure functions/methods including edge cases work as intended?

---

This is a new function that will be used across the entire REST API. Tests could cover:

```
# Test cases needed:
- param_get_bool with 'true' → returns True
- param_get_bool with 'false' → returns False  
- param_get_bool with '1' → returns True + logs warning
- param_get_bool with '0' → returns False + logs warning
- param_get_bool with True (already bool) → returns True (no warning)
- param_get_bool with False (already bool) → returns False (no warning)
- param_get_bool with 'invalid' → raises 400 error with TypeError
- param_get_bool with None and no default → raises 400 error with KeyError
- param_get_bool with None and default=False → returns False
- param_get_bool with missing param and default=True → returns True
- param_get_bool with case variations ('TRUE', 'False', 'TrUe') → handles correctly
```

---

Re: test structure - I'd suggest creating `tests/unit/rucio/web/rest/flaskapi/v1/test_common.py` to mirror the source structure (similar to how you already have tests/rucio/common/test_utils.py). This keeps unit tests organized and makes them easy to find.

For a broader cleanup, consider gradually moving toward separating `tests/unit/` and `tests/integration/` directories (ref: https://github.com/rucio/rucio/issues/7946#issuecomment-3275318572). But for this PR, just adding the REST unit tests in a mirrored structure would be a good incremental step. See the test structure proposal above for pros/cons of different approaches if you want to tackle this more comprehensively later.

**Review by mgajek-cern — COMMENTED** [review]
Just a few findings/questions/remarks below:

File: lib/rucio/web/rest/flaskapi/v1/common.py (hunk 2 above)
Comment (voetberg) [reply lib/rucio/web/rest/flaskapi/v1/common.py]: I'll add tests but based on the above issue it looks like it still needs some discussion. I'll hold off completely reorganizing tests in this issue.

Comment (voetberg) [reply lib/rucio/web/rest/flaskapi/v1/common.py]: Addressed a few of those. Others are not strictly boolean options

File: tests/test_utils_rest.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+import pytest
+from werkzeug import exceptions
+
+from rucio.web.rest.flaskapi.v1.common import param_get_bool
+
+
+@pytest.mark.parametrize(
+    'params,kwargs,raise_log,raise_error',
+    [
+        ({"object": False}, {}, False, False),
```
Comment (mgajek-cern) [inline tests/test_utils_rest.py:25]: Can we add extra cases, e.g.

```python
({"object": True}, {}, False, False),            # Pure True value - no warning
({}, {"default": False}, False, False),          # default=False works
({"object": '1'}, {}, True, False),             # String '1' - logs warning, returns True
({"object": None}, {"default": False}, False, False),  # Explicit None with default
```

Additional case variations:

```python
({"object": "TRUE"}, {}, False, False),
({"object": "False"}, {}, False, False), 
({"object": "FALSE"}, {}, False, False),
```

Cases to be cautious about (need manual testing first):

```python
({"object": 0}, {}, True, False)  # Integer 0 - MIGHT trigger warning due to int(0) succeeding
```

File: lib/rucio/web/rest/flaskapi/v1/common.py (hunk 2 above)
Comment (mgajek-cern) [reply lib/rucio/web/rest/flaskapi/v1/common.py]: I went through the list of `param_get(..., default=False/True)` occurrences and checked how the parameters are used. Most of them (like inherit, force, ignore_availability, locked, dry_run, etc.) are indeed passed into logic that expects a boolean value, e.g.as flags for downstream functions or used in conditionals.

Could you confirm which of the listed parameters are not strictly boolean options, so we can safely convert the remaining ones to use `param_get_bool`?

**Review by mgajek-cern — COMMENTED** [review]
A few follow up questions/remarks:

File: tests/test_utils_rest.py (hunk 3 above)
Comment (voetberg) [reply tests/test_utils_rest.py:25]: The explicit None should raise an error (set an invalid type parameter) but otherwise adding the additional tests

File: lib/rucio/web/rest/flaskapi/v1/common.py (hunk 2 above)
Comment (voetberg) [reply lib/rucio/web/rest/flaskapi/v1/common.py]: Rule get - estimate_ttc
subscriptions post - priority
 
Everything else can be converted

**Review by mgajek-cern — COMMENTED** [review]
The `0% Codecov patch coverage` might be due to a CI configuration (see: https://github.com/rucio/rucio/issues/8067).

Additional considerations for follow-up:

- Type hints: Adding proper type annotations could help satisfy `python_annotations` job in the `autotest.yml` workflow
- VO-tests failure: One job appears to be failing in the VO-tests, worth investigating if it's related to this boolean parameter handling or an unrelated issue

The core functionality is well-tested and the coverage issue should be addressed separately in the CI configuration.

Ready to approve after these are addressed, though another reviewer is welcome to check too

**Comment by voetberg** [comment]
> Additional considerations for follow-up:
> 
>     * Type hints: Adding proper type annotations could help satisfy `python_annotations` job in the `autotest.yml` workflow

This is because "kwargs" is being picked up as an arg to be annotated, which I fundamentally disagree with

>     * VO-tests failure: One job appears to be failing in the VO-tests, worth investigating if it's related to this boolean parameter handling or an unrelated issue
 This is a rule lock problem on - tests/test_judge_evaluator.py::test_judge_double_container_with_existing_rule - a known flaky test updated by #8117 after this PR went in

**Review by mgajek-cern — DISMISSED** [review]

File: lib/rucio/web/rest/flaskapi/v1/common.py [hunk 4]
```diff
@@ -425,6 +425,48 @@ def param_get(parameters: dict[str, Any], name: str, **kwargs) -> Any:
...
+    """
+        Get a boolean parameter from the passed parameters. Converts to True/False.
+    """
+    def convert_str(option: Union[str, bool]) -> bool:
+        # TODO remove warning and replace with error in v40 - #PLACEHOLDER ISSUE
```
Comment (rdimaio) [inline lib/rucio/web/rest/flaskapi/v1/common.py]: imo we should already create an issue and put it in this comment for reference

File: lib/rucio/web/rest/flaskapi/v1/common.py [hunk 5]
```diff
@@ -425,6 +425,48 @@ def param_get(parameters: dict[str, Any], name: str, **kwargs) -> Any:
         return parameters[name]
 
 
+def param_get_bool(parameters: dict[str, Any], name: str, **kwargs) -> bool:
+    """
+        Get a boolean parameter from the passed parameters. Converts to True/False.
+    """
+    def convert_str(option: Union[str, bool]) -> bool:
```
Comment (rdimaio) [inline lib/rucio/web/rest/flaskapi/v1/common.py]: For clarity I would call this `convert_str_to_bool`, or maybe just `str_to_bool`. IMO I would also add an underscore at the beginning to signify that it's only used internally within the context of `param_get_bool`

File: lib/rucio/web/rest/flaskapi/v1/accounts.py [hunk 6]
```diff
@@ -654,7 +654,7 @@ def post(self, account: str) -> 'ResponseReturnValue':
                     default: none
                   default:
                     description: "Should this be the default account?"
-                    type: string
+                    type: boolean
```
Comment (rdimaio) [inline lib/rucio/web/rest/flaskapi/v1/accounts.py:657]: nit: should be in a separate commit

File: lib/rucio/web/rest/flaskapi/v1/common.py [hunk 7]
```diff
@@ -425,6 +425,48 @@ def param_get(parameters: dict[str, Any], name: str, **kwargs) -> Any:
         return parameters[name]
 
 
+def param_get_bool(parameters: dict[str, Any], name: str, **kwargs) -> bool:
```
Comment (rdimaio) [inline lib/rucio/web/rest/flaskapi/v1/common.py]: Instead of `kwargs`, I would propose just having `default` as a defined parameter, as that's the only extra `kwarg` that I see being used

Comment (rdimaio) [reply lib/rucio/web/rest/flaskapi/v1/common.py]: (and really, same goes for the original `param_get` function - we pass `kwargs but we only use `default`)

Comment (voetberg) [reply lib/rucio/web/rest/flaskapi/v1/common.py]: I was thinking that as well, I was debating on changing both or just using the `param_get` convention. I chose wrong ig lol

**Review by rdimaio — APPROVED** [review]
